test/xds/xds_server_test.go

```diff
@@ -0,0 +1,428 @@
+		t.Fatalf("Failed to create an xDS enabled gRPC server: %v", err)
+	}
+	defer server.Stop()
+	testgrpc.RegisterTestServiceServer(server, &testService{})
+	go func() {
+		if err := server.Serve(lis); err != nil {
+			t.Errorf("Serve() failed: %v", err)
+		}
+	}()
+	<-serving.Done()
+
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	waitForSuccessfulRPC(ctx, t, cc) // Eventually, the LDS and dynamic RDS get processed, work, and RPC's should work as usual.
+
+	// Set the route config to be of type route action route, which the rpc will
+	// match to. This should eventually reflect in the Conn's routing
+	// configuration and fail the rpc with a status code UNAVAILABLE.
+	routeConfig = e2e.RouteConfigFilterAction("routeName")
+	resources = e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: []*v3listenerpb.Listener{listener}, // Same lis, so will get eaten by the xDS Client.
+
+		Routes: []*v3routepb.RouteConfiguration{routeConfig},
+	}
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// "NonForwardingAction is expected for all Routes used on server-side; a
+	// route with an inappropriate action causes RPCs matching that route to
+	// fail with UNAVAILABLE." - A36
+	waitForFailedRPCWithStatusCode(ctx, t, cc, status.New(codes.Unavailable, "the incoming RPC matched to a route that was not of action type non forwarding"))
+}
+
+// waitForFailedRPCWithStatusCode makes Unary RPC's until it receives the
+// expected status in a polling manner. Fails if the RPC made does not return
+// the expected status before the context expires.
+func waitForFailedRPCWithStatusCode(ctx context.Context, t *testing.T, cc *grpc.ClientConn, sts ...*status.Status) {
+	t.Helper()
+
+	c := testgrpc.NewTestServiceClient(cc)
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+	var err error
+	for {
+		select {
+		case <-ctx.Done():
+			t.Fatalf("failure when waiting for RPCs to fail with certain status %v: %v. most recent error received from RPC: %v", sts, ctx.Err(), err.Error())
+		case <-ticker.C:
+			_, err = c.EmptyCall(ctx, &testpb.Empty{})
+			for _, st := range sts {
+				if status.Code(err) == st.Code() && strings.Contains(err.Error(), st.Message()) {
+					t.Logf("most recent error happy case: %v", err.Error())
+					return
+				}
+			}
+		}
+	}
+}
+
+// TestResourceNotFoundRDS tests the case where an LDS points to an RDS which
+// returns resource not found. Before getting the resource not found, the xDS
+// Server has not received all configuration needed, so it should Accept and
+// Close any new connections. After it has received the resource not found
+// error, the server should move to serving, successfully Accept Connections,
+// and fail at the L7 level with resource not found specified.
+func (s) TestResourceNotFoundRDS(t *testing.T) {
```

Comment:
Looks like a lot of WIP tests from this point on. So, I'm stopping here.

Reply:
Added.